### PR TITLE
Introduction of pupil restart reason table and variety of fixes to maintain functionality

### PR DESCRIPTION
**Pull request by GeorgeChatzigiannis**
- New migrations for renaming statusDesc to description in all tables, restart reasons table
- Fixes to maintain functionality

### CODE REVIEW

File: admin/services/data-access/pupil-restart.data.service.js [hunk 1]
```diff
@@ -115,20 +115,80 @@ pupilRestartDataService.sqlFindLatestRestart = async function (pupilId) {
...
   return sqlService.query(sql)
 }
 
+/**
+ * Find pupil's restart reason description by id
+ * @param id
+ * @return {Promise.<void>}
+ */
+pupilRestartDataService.sqlFindRestartReasonDescById = async function (id) {
+  console.log('fdsfdsdsfdsfdsfasd', id)
```
Comment (GuyHarwood) [inline admin/services/data-access/pupil-restart.data.service.js]: please remove this, or change the message and use winston

File: admin/services/data-access/pupil-restart.data.service.js [hunk 2]
```diff
@@ -115,20 +115,80 @@ pupilRestartDataService.sqlFindLatestRestart = async function (pupilId) {
...
+  const params = [
+    {
+      name: 'code',
+      value: code,
+      type: TYPES.Char
+    }
+  ]
+  const result = await sqlService.query(sql, params)
+  const obj = R.head(result)
+  return obj.id
```
Comment (GuyHarwood) [inline admin/services/data-access/pupil-restart.data.service.js]: what if it's empty or undefined?

File: admin/services/data-access/pupil-restart.data.service.js [hunk 3]
```diff
@@ -115,20 +115,80 @@ pupilRestartDataService.sqlFindLatestRestart = async function (pupilId) {
...
+  const params = [
+    {
+      name: 'id',
+      value: id,
+      type: TYPES.Int
+    }
+  ]
+  const result = await sqlService.query(sql, params)
+  const obj = R.head(result)
+  return obj['description']
```
Comment (GuyHarwood) [inline admin/services/data-access/pupil-restart.data.service.js]: what if it's empty or undefined?

File: admin/services/data-access/pupil-restart.data.service.js [hunk 4]
```diff
@@ -149,8 +209,8 @@ pupilRestartDataService.sqlMarkRestartAsDeleted = async (pupilId, userId) => {
     }
   ]
   const sql = `UPDATE ${sqlService.adminSchema}.[pupilRestart] 
-  SET isDeleted=1, deletedByUser_id=@userId  
-  WHERE [pupil_id]=@pupilId`
+  SET isDeleted=1, deletedByUser_id=@deletedByUser_id  
+  WHERE pupil_id = @pupil_id`
```
Comment (GuyHarwood) [inline admin/services/data-access/pupil-restart.data.service.js]: the parameter names don't _need_ the underscore

File: admin/services/restart.service.js [hunk 5]
```diff
@@ -129,21 +138,22 @@ restartService.getSubmittedRestarts = async schoolId => {
...
-    const record = latestPupilRestarts.find(l => l.pupilId === p.id)
+  await Promise.all(pupils.map(async p => {
+    const record = latestPupilRestarts.find(l => l.pupil_id === p.id)
     if (record) {
+      const reason = await pupilRestartDataService.sqlFindRestartReasonDescById(record.pupilRestartReason_id)
       restarts.push({
         id: record.id,
         pupilId: p.id,
-        reason: record.reason,
+        reason,
```
Comment (GuyHarwood) [inline admin/services/restart.service.js]: please just add the property name, its far more readable and consistent with the other properties

**Comment by GuyHarwood** [comment]
@GeorgeChatzigiannis running your branch locally i get 2 failing tests...

```
​​​​​2 failing tests, 724 passing​​​​​
  
  ​​​​restart controller: getSelectRestartList route displays the restart pupils list page​​​​ ​​​[5006 ms]​​​​
    ​​​​
    ​​Error: Timeout - Async callback was not invoked within timeout specified by jasmine.DEFAULT_TIMEOUT_INTERVAL.​​
    ​​​​​​​​​​
    ​​​​​From console:​​​​​
      ​​​​​​​​​​
      ​​​​​error:  ConnectionError: Failed to connect to localhost:1433 - Could not connect (sequence)​​​​​
      ​​​​​    at ConnectionError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/errors.js:12:12)​​​​​
      ​​​​​    at Connection.socketError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connection.js:684:30)​​​​​
      ​​​​​    at /Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connection.js:575:25​​​​​
      ​​​​​    at SequentialConnectionStrategy.connect (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connector.js:153:9)​​​​​
      ​​​​​    at Socket.onError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connector.js:169:16)​​​​​
      ​​​​​    at emitOne (events.js:115:13)​​​​​
      ​​​​​    at Socket.emit (events.js:210:7)​​​​​
      ​​​​​    at emitErrorNT (internal/streams/destroy.js:64:8)​​​​​
      ​​​​​    at _combinedTickCallback (internal/process/next_tick.js:138:11)​​​​​
      ​​​​​    at process._tickCallback (internal/process/next_tick.js:180:9)​​​​​
      ​​​​​​​​​​
      ​​​​​​​​​​
      ​​​​​error:  ConnectionError: Failed to connect to localhost:1433 - Could not connect (sequence)​​​​​
      ​​​​​    at ConnectionError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/errors.js:12:12)​​​​​
      ​​​​​    at Connection.socketError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connection.js:684:30)​​​​​
      ​​​​​    at /Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connection.js:575:25​​​​​
      ​​​​​    at SequentialConnectionStrategy.connect (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connector.js:153:9)​​​​​
      ​​​​​    at Socket.onError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connector.js:169:16)​​​​​
      ​​​​​    at emitOne (events.js:115:13)​​​​​
      ​​​​​    at Socket.emit (events.js:210:7)​​​​​
      ​​​​​    at emitErrorNT (internal/streams/destroy.js:64:8)​​​​​
      ​​​​​    at _combinedTickCallback (internal/process/next_tick.js:138:11)​​​​​
      ​​​​​    at process._tickCallback (internal/process/next_tick.js:180:9)​​​​​
      ​​​​​​​​​​
      ​​​​​​​​​​
      ​​​​​error:  ConnectionError: Failed to connect to localhost:1433 - Could not connect (sequence)​​​​​
      ​​​​​    at ConnectionError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/errors.js:12:12)​​​​​
      ​​​​​    at Connection.socketError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connection.js:684:30)​​​​​
      ​​​​​    at /Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connection.js:575:25​​​​​
      ​​​​​    at SequentialConnectionStrategy.connect (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connector.js:153:9)​​​​​
      ​​​​​    at Socket.onError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connector.js:169:16)​​​​​
      ​​​​​    at emitOne (events.js:115:13)​​​​​
      ​​​​​    at Socket.emit (events.js:210:7)​​​​​
      ​​​​​    at emitErrorNT (internal/streams/destroy.js:64:8)​​​​​
      ​​​​​    at _combinedTickCallback (internal/process/next_tick.js:138:11)​​​​​
      ​​​​​    at process._tickCallback (internal/process/next_tick.js:180:9)​​​​​
      ​​​​​​​​​​
      ​​​​​​​​​​
      ​​​​​error:  ConnectionError: Failed to connect to localhost:1433 - Could not connect (sequence)​​​​​
      ​​​​​    at ConnectionError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/errors.js:12:12)​​​​​
      ​​​​​    at Connection.socketError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connection.js:684:30)​​​​​
      ​​​​​    at /Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connection.js:575:25​​​​​
      ​​​​​    at SequentialConnectionStrategy.connect (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connector.js:153:9)​​​​​
      ​​​​​    at Socket.onError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connector.js:169:16)​​​​​
      ​​​​​    at emitOne (events.js:115:13)​​​​​
      ​​​​​    at Socket.emit (events.js:210:7)​​​​​
      ​​​​​    at emitErrorNT (internal/streams/destroy.js:64:8)​​​​​
      ​​​​​    at _combinedTickCallback (internal/process/next_tick.js:138:11)​​​​​
      ​​​​​    at process._tickCallback (internal/process/next_tick.js:180:9)​​​​​
      ​​​​​​​​​​
      ​​​​​​​​​​
      ​​​​​error:  ConnectionError: Failed to connect to localhost:1433 - Could not connect (sequence)​​​​​
      ​​​​​    at ConnectionError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/errors.js:12:12)​​​​​
      ​​​​​    at Connection.socketError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connection.js:684:30)​​​​​
      ​​​​​    at /Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connection.js:575:25​​​​​
      ​​​​​    at SequentialConnectionStrategy.connect (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connector.js:153:9)​​​​​
      ​​​​​    at Socket.onError (/Users/guyharwood/dev/MTC/admin/node_modules/tedious-connection-pool/node_modules/tedious/lib/connector.js:169:16)​​​​​
      ​​​​​    at emitOne (events.js:115:13)​​​​​
      ​​​​​    at Socket.emit (events.js:210:7)​​​​​
      ​​​​​    at emitErrorNT (internal/streams/destroy.js:64:8)​​​​​
      ​​​​​    at _combinedTickCallback (internal/process/next_tick.js:138:11)​​​​​
      ​​​​​    at process._tickCallback (internal/process/next_tick.js:180:9)​​​​​
      ​​​​​​​​​​
  
  ​​​​check.data.service #sqlFindNumberOfChecksStartedByPupil makes the expected calls​​​​ ​​​[6 ms]​​​​
    ​​​​
    ​​TypeError: Cannot read property 'cnt' of undefined​​
      at Object.checkDataService.sqlFindNumberOfChecksStartedByPupil ​services/data-access/check.data.service.js:154:2​
```

File: admin/services/data-access/pupil-restart.data.service.js (hunk 1 above)
Comment (GeorgeChatzigiannis) [reply admin/services/data-access/pupil-restart.data.service.js]: @GuyHarwood removed

File: admin/services/data-access/pupil-restart.data.service.js (hunk 3 above)
Comment (GeorgeChatzigiannis) [reply admin/services/data-access/pupil-restart.data.service.js]: @GuyHarwood updated to verify if object is defined first

File: admin/services/data-access/pupil-restart.data.service.js (hunk 2 above)
Comment (GeorgeChatzigiannis) [reply admin/services/data-access/pupil-restart.data.service.js]: @GuyHarwood updated to verify if object is defined first

File: admin/services/data-access/pupil-restart.data.service.js (hunk 4 above)
Comment (GeorgeChatzigiannis) [reply admin/services/data-access/pupil-restart.data.service.js]: @GuyHarwood renamed both

File: admin/services/restart.service.js (hunk 5 above)
Comment (GeorgeChatzigiannis) [reply admin/services/restart.service.js]: @GuyHarwood updated

File: admin/lib/validator/restart-validator.js [hunk 6]
```diff
@@ -10,9 +10,10 @@ const restartValidator = {}
  */
 restartValidator.validateReason = (reason, specifyReason) => {
   const validationError = new ValidationError()
-  if (reason === 'Did not complete' && specifyReason.length === 0) {
+  if (reason === 'DNC' && specifyReason.length === 0) {
     validationError.addError('didNotCompleteInfo', restartErrorMessages.didNotCompleteInfo)
   }
+  console.log('validationError', validationError)
```
Comment (jon-shipley) [inline admin/lib/validator/restart-validator.js]: debug? remove?

File: admin/services/data-access/pupil-restart.data.service.js [hunk 7]
```diff
@@ -115,20 +116,79 @@ pupilRestartDataService.sqlFindLatestRestart = async function (pupilId) {
...
+  const params = [
+    {
+      name: 'id',
+      value: id,
+      type: TYPES.Int
+    }
+  ]
+  const result = await sqlService.query(sql, params)
+  const obj = R.head(result)
+  return obj && obj['description']
```
Comment (jon-shipley) [inline admin/services/data-access/pupil-restart.data.service.js]: FYI - `R.prop('description', obj)` would return the property or undefined if the property or the object did not exist.
http://ramdajs.com/docs/#prop

File: admin/services/data-access/pupil-restart.data.service.js [hunk 8]
```diff
@@ -115,20 +116,79 @@ pupilRestartDataService.sqlFindLatestRestart = async function (pupilId) {
...
+ * @return {Promise.<void>}
+ */
+pupilRestartDataService.sqlFindRestartReasons = async function () {
+  const sql = `
+  SELECT 
+    id, 
+    code, 
+    description
+  FROM ${sqlService.adminSchema}.[pupilRestartReason]
+  ORDER BY id ASC`
```
Comment (jon-shipley) [inline admin/services/data-access/pupil-restart.data.service.js]: Ordering by `id` seems odd. If the GUI wants the data to be ordered then add in a `displayOrder` field and order by that.  Ordering by `id` is a bad practise.

File: admin/lib/validator/restart-validator.js (hunk 6 above)
Comment (GeorgeChatzigiannis) [reply admin/lib/validator/restart-validator.js]: @jon-shipley removed

File: admin/services/data-access/pupil-restart.data.service.js (hunk 7 above)
Comment (GeorgeChatzigiannis) [reply admin/services/data-access/pupil-restart.data.service.js]: @jon-shipley yea i completely forgot about this, it's added

File: admin/services/data-access/pupil-restart.data.service.js (hunk 8 above)
Comment (GeorgeChatzigiannis) [reply admin/services/data-access/pupil-restart.data.service.js]: @jon-shipley i need to sort the reason based on the way they have to appear in the front-end. if i don't order by id they will appear at a random order.
I am guessing the only way to solve this is to add an order field?

Comment (jon-shipley) [reply admin/services/data-access/pupil-restart.data.service.js]: I'd call it `displayOrder` or `guiOrder` to show that it is a display related field.  Please don't call it `order` as that is a reserved word (and `attendanceCode.order` is now tech debt.

Comment (GeorgeChatzigiannis) [reply admin/services/data-access/pupil-restart.data.service.js]: @jon-shipley updated using `displayOrder`

**Pull request merged**